crates/biome_js_analyze/src/lint/nursery/no_async_client_component.rs

```diff
@@ -0,0 +1,139 @@
+use crate::react::components::{
+    AnyPotentialReactComponentDeclaration, ReactComponentInfo, ReactComponentKind,
+};
+use biome_analyze::{
+    Ast, Rule, RuleDiagnostic, RuleDomain, RuleSource, context::RuleContext, declare_lint_rule,
+};
+use biome_console::markup;
+use biome_diagnostics::Severity;
+use biome_js_syntax::AnyJsRoot;
+use biome_rowan::{AstNode, AstNodeList};
+use biome_rule_options::no_async_client_component::NoAsyncClientComponentOptions;
+
+declare_lint_rule! {
+    /// Prevent client components from being async functions.
+    ///
+    /// This rule prevents the use of async functions for client components in Next.js applications.
+    /// Client components marked with "use client" directive should not be async as this can cause
+    /// hydration mismatches, break component rendering lifecycle, and lead to unexpected behavior
+    /// with React's concurrent features.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```jsx,expect_diagnostic
+    /// "use client";
+    ///
+    /// export default async function MyComponent() {
+    ///   return <div>Hello</div>;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```jsx
+    /// "use client";
+    ///
+    /// export default function MyComponent() {
+    ///   return <div>Hello</div>;
+    /// }
+    /// ```
+    ///
+    /// ```jsx
+    /// // No "use client" directive - server component can be async
+    /// export default async function ServerComponent() {
+    ///   const data = await fetch('/api/data');
+    ///   return <div>{data}</div>;
+    /// }
+    /// ```
+    ///
+    pub NoAsyncClientComponent {
+        version: "next",
+        name: "noAsyncClientComponent",
```

Comment:
This purely applies to just Next.js applications right? We should name this `noNextAsyncClientComponent`.

Reply:
The files, and the diagnostic category too